src/app/utils/pickupTimeEstimator.js

```diff
@@ -0,0 +1,154 @@
+	if (roundedHours !== 0 && roundedMinutes !== 0) str += ' '
+	if (minutes !== 0) str += `${roundedMinutes} minutes`
+	return str
+}
+
+export const _roundToQuarterHour = (hours, minutes) => {
+	if (minutes < 60 && minutes > 45) {
+		hours++
+		minutes = 0
+	}
+	if (minutes % 15 !== 0) {
+		minutes += 15 - (minutes % 15)
+	}
+	return { hours, minutes }
+}
+
+// expects a timestamp (integer), returns a string
+export const _buildTimeString = (estimatedDeliveryTime) => {
+	let minutes = estimatedDeliveryTime.getMinutes()
+	let hours = estimatedDeliveryTime.getHours()
+	// round up to the nearest quarter hour, with special accounting for the 
+	// last quarter before the hour, because the hour needs to increment
+	let { hours: roundedHours, minutes: roundedMinutes } = _roundToQuarterHour(hours, minutes)
+
+	if (roundedMinutes < 10) roundedMinutes = `0${roundedMinutes}`
+	const amOrPm = roundedHours > 11 ? 'pm' : 'am'
+	roundedHours = roundedHours % 12 === 0 ? 12 : roundedHours % 12
+	return `${roundedHours}:${roundedMinutes} ${amOrPm}`
+}
+
+// Delivery location should be sc, ma, or my (2 letter codes for each research branch). Request time is a Date object
+export const _expectedAvailableDay = async (deliveryLocation, requestTime, duration) => {
+	const locationHours = await _operatingHours(deliveryLocation)
+	let available
+	let today = new Date(locationHours[0].startTime).getDay()
+	let provisionalDeliveryTime
+	const hours = locationHours.find((day, i) => {
+		// convert everything into ms:
+		const { endTime } = day
+		const endTimeInMs = Date.parse(endTime)
+		provisionalDeliveryTime = Date.parse(requestTime) + duration
+		const requestTimeMs = Date.parse(requestTime)
+		const finalRequestTimeMs = endTimeInMs - REQUEST_CUTOFF_BUFFER
+		//
+
+		// if request was made after request cutoff time, today is not your day
+		if (requestTimeMs > finalRequestTimeMs) return false
+		// if estimated delivery time is before the end of the current day, current
+		// day is the day.
+		if (provisionalDeliveryTime < endTimeInMs) {
+			// determine if that is today, tomorrow, or two days from now.
+			const nextBusinessDay = new Date(provisionalDeliveryTime).getDay()
+			available = _determineNextBusinessDay(today, nextBusinessDay, i)
+			return true
+		}
+	})
+
+	return {
+		available,
+		estimatedDeliveryTime: _calculateDeliveryTime(available, provisionalDeliveryTime, hours.startTime),
+		day: hours.day
+	}
+}
+
+// If it can be delivered today, the provisional estimation still serves.
+// Otherwise, estimated delivery time is the startTime plus the amount
+// of time the branch needs to get their ducks in a row.
+export const _calculateDeliveryTime = (dayAvailable, provisionalDeliveryTime, startTime) => {
+	if (dayAvailable === 'today') return new Date(provisionalDeliveryTime)
+	else {
+		return new Date(Date.parse(startTime) + OPENING_BUFFER)
+	}
+}
+
+export const _determineNextBusinessDay = (today, nextBusinessDay, i) => {
```

Comment:
This method takes two Date objects and an "i", which represents, in effect, the number of days between the date objects?

Reply:
ill update the code, but all three are numbers. today and nextbusinessday are numbers that correspond to a day of the week starting with sunday = 0. today is the day the code is executed. next business day is the next day the book can be delivered. i is the index of the hours array from the locations service. we only care about it if its 0, because it indicates that the item is deliverable today.

there is a potential bug with the code. this function can't tell the difference between tomorrow (thursday) and thursday next week. this would only be a problem if the library is closed for more than a few days in a row/a week. should we worry about that case?

Reply:
i refactored this function to only use today and nextDeliverableDay, and added a comment describing the function. also found and squashed a bug because of that!